packages/blade/src/components/ActionList/ActionList.tsx

```diff
@@ -1,119 +1,149 @@
-};
-const ActionList = ({ children }: ActionListProps): JSX.Element => {
-  const { setOptions, actionListRef, selectionType, dropdownBaseId } = useDropdown();
-  const actionListOptions: {
-    title: string;
-    value: string;
-  }[] = [];
+      const clonedChild = React.cloneElement(child, {
+        // @ts-expect-error: TS doesn't understand the child's props
+        _index: currentIndex,
+      });
+      return clonedChild;
+    }
 
-  console.count('ActionList');
+    return child;
+  };
+
+  const childCompIdArray = React.Children.toArray(children).map((child) => getComponentId(child));
+  const lastActionListSectionIndex = childCompIdArray.lastIndexOf(componentIds.ActionListSection);
+
+  const isActionListItemPresentAfterSection = childCompIdArray
+    .slice(lastActionListSectionIndex)
+    .includes(componentIds.ActionListItem);
 
   // Looping through ActionListItems to add index to them and get an options array for moving focus between items
-  const childrenWithId = React.Children.map(children, (child) => {
+  const childrenWithId = React.Children.map(children, (child, index) => {
     if (React.isValidElement(child)) {
-      // @TODO: handle the scenario where ActionListItem is inside ActionListMenu
-      if (child.type === ActionListItem) {
-        actionListOptions.push({
-          title: child.props.title,
-          value: child.props.value,
-        });
-        const currentIndex = actionListOptions.length - 1;
-        const clonedChild = React.cloneElement(child, {
+      if (isValidAllowedChildren(child, componentIds.ActionListSection)) {
+        return React.cloneElement(child, {
           // @ts-expect-error: TS doesn't understand the child's props
-          _index: currentIndex,
+          children: React.Children.map(child.props.children, (childInSection) => {
+            if (isValidAllowedChildren(childInSection, componentIds.ActionListItem)) {
+              return getActionListItemChildWithId(childInSection);
+            }
+
+            return childInSection;
+          }),
+          hideDivider: index === lastActionListSectionIndex && !isActionListItemPresentAfterSection,
         });
-        return clonedChild;
       }
-    }
 
+      if (isValidAllowedChildren(child, componentIds.ActionListItem)) {
+        return getActionListItemChildWithId(child);
+      }
+    }
     return child;
   });
 
   React.useEffect(() => {
     setOptions(actionListOptions);
     // eslint-disable-next-line react-hooks/exhaustive-deps
-  }, [setOptions, children]);
+  }, [optionsRecalculateToggle]);
+
+  React.useEffect(() => {
+    setSelectedIndices(defaultSelectedIndices);
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, []);
 
   return (
-    <Box
+    <StyledActionList
       // eslint-disable-next-line @typescript-eslint/no-explicit-any
       ref={actionListRef as any}
       {...makeAccessible({
-        role: 'listbox',
+        role: isReactNative() ? 'menu' : 'listbox',
```

Comment:
so I had tried that earlier but screenreader announces role="dialog" explicitly as dialog which I though would throw users away since dialogs are mostly modals.

So now, I am conditionally checking if footer has buttons. If it has buttons, then I switch to `dialog` role but otherwise I stay on `listbox` only. https://github.com/razorpay/blade/pull/966/commits/f1560106cb47c0d7576f345327f3d38174f7717a